bfvmm/src/hve/arch/intel_x64/xen/xen_op.cpp

```diff
@@ -1147,6 +1170,39 @@ xen_op_handler::VCPUOP_stop_periodic_timer_handler(gsl::not_null<vcpu *> vcpu)
     vcpu->set_rax(SUCCESS);
 }
 
+void
+xen_op_handler::VCPUOP_stop_singleshot_timer_handler(gsl::not_null<vcpu *> vcpu)
+{
+    vcpu->set_rax(SUCCESS);
+}
+
+void
+xen_op_handler::VCPUOP_set_singleshot_timer_handler(gsl::not_null<vcpu *> vcpu)
+{
+    try {
+        expects(vcpu->rsi() == 0);
+
+        auto arg = vcpu->map_arg<vcpu_set_singleshot_timer_t>(vcpu->rdx());
+        auto deadline = arg->timeout_abs_ns;
+        auto now = this->tsc_to_sys_time();
+
+        if ((arg->flags & VCPU_SSHOTTMR_future) && deadline < now) {
+            vcpu->set_rax(FAILURE);
+            return;
+        }
+
+        auto tsc_ticks = now - deadline;
+        auto pet_ticks = tsc_ticks >> m_pet_shift;
+
+        m_vcpu->set_vmx_preemption_timer(pet_ticks);
```

Comment:
Don't forget that when you exit the VM for an external interrupt, you need to also wait in bfexec. This is a REALLY good sign BTW.

Reply:
Hopefully you see a yield soon